tutorial/python_agents/base_agent.py

```diff
@@ -0,0 +1,475 @@
+#        distribution.
+#
+#      * Neither the name of Intel Corporation nor the names of its
+#        contributors may be used to endorse or promote products derived
+#        from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+#  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+'''
+This package provides a BaseAgent class for monitoring jobs. One can also
+extends this class to implement more functionality around the convenience
+provided by it.
+
+In addition, the package provides parse_common_args and run_controller
+convenience methods for easily creating command line for classes that extend
+BaseAgent.
+'''
+
+import sys
+import os
+import argparse
+from collections import OrderedDict
+import socket
+import yaml
+
+import geopmdpy
+import geopmdpy.runtime
+import geopmpy.reporter
+
+
+CSV_SEP_CHAR = "|"
+
+DOMAIN_SEP_CHAR = "-"
+
+TIME_SIGNAL = ("TIME", geopmdpy.topo.DOMAIN_BOARD, 0, "-")
+
+DEFAULT_SIGNAL_LIST = [
+    ("INSTRUCTIONS_RETIRED",         geopmdpy.topo.DOMAIN_BOARD,        0, "incr"),
+    ("POWER_PACKAGE",                geopmdpy.topo.DOMAIN_BOARD,        0, "avg"),
+    ("POWER_DRAM",                   geopmdpy.topo.DOMAIN_BOARD_MEMORY, 0, "avg"),
+    ("FREQUENCY",                    geopmdpy.topo.DOMAIN_BOARD,        0, "avg"),
+    ("MSR::UNCORE_PERF_STATUS:FREQ", geopmdpy.topo.DOMAIN_BOARD,        0, "avg"),
+    ("TEMPERATURE_CORE",             geopmdpy.topo.DOMAIN_BOARD,        0, "avg"),
+    ("ENERGY_DRAM",                  geopmdpy.topo.DOMAIN_BOARD_MEMORY, 0, "incr"),
+    ("ENERGY_PACKAGE",               geopmdpy.topo.DOMAIN_BOARD,        0, "incr"),
+]
+
+
+class BaseAgent(geopmdpy.runtime.Agent):
+    ''' Simple GEOPM python agent that only monitors jobs. It is similar to but not an
+    equivalent of the GEOPM C++ monitor agent. In the general scheme, one can use this agent to
+    monitor a running job or use it as a base class (or template) for agents with more
+    functionality.
+
+    BaseAgent has the capability to print a trace output of the monitored signals in a CSV file with
+    base_agent.CSV_SEP_CHAR as the field separator. However it has 2 differences in the trace
+    that are meant to be conveniences:
```

Comment:
No longer diffing values in the trace, though the diffs are provided to the agent for ease of use.